src/job_cache/job_cache_impl_common.cpp

```diff
@@ -0,0 +1,251 @@
+  if (fstat(src_fd, &buf) < 0) {
+    log_fatal("fstat(src_fd = %d): %s", src_fd, strerror(errno));
+  }
+  off_t idx = 0;
+  size_t size = buf.st_size;
+  do {
+    intptr_t written = sendfile(dst_fd, src_fd, &idx, size);
+    if (written < 0) {
+      log_fatal("sendfile(src_fd = %d, NULL, dst_fd = %d, size = %d, 0): %s", src_fd, dst_fd,
+                buf.st_size, strerror(errno));
+    }
+    idx = written;
+    size -= written;
+  } while (size != 0);
+}
+#endif
+
+#ifdef FICLONE
+
+void copy_or_reflink(const char *src, const char *dst, mode_t mode, int extra_flags) {
+  auto src_fd = wcl::unique_fd::open(src, O_RDONLY);
+  if (!src_fd) {
+    log_fatal("open(%s): %s", src, strerror(src_fd.error()));
+  }
+  auto dst_fd = wcl::unique_fd::open(dst, O_WRONLY | O_CREAT | extra_flags, mode);
+  if (!dst_fd) {
+    log_fatal("open(%s): %s", dst, strerror(dst_fd.error()));
+  }
+
+  if (ioctl(dst_fd->get(), FICLONE, src_fd->get()) < 0) {
+    if (errno != EINVAL && errno != EOPNOTSUPP && errno != EXDEV) {
+      log_fatal("ioctl(%s, FICLONE, %d): %s", dst, src, strerror(errno));
+    }
+    copy(src_fd->get(), dst_fd->get());
+  }
+}
+
+#else
+
+void copy_or_reflink(const char *src, const char *dst, mode_t mode, int extra_flags) {
+  auto src_fd = wcl::unique_fd::open(src, O_RDONLY);
+  if (!src_fd) {
+    log_fatal("open(%s): %s", src, strerror(src_fd.error()));
+  }
+  auto dst_fd = wcl::unique_fd::open(dst, O_WRONLY | O_CREAT | extra_flags, mode);
+  if (!dst_fd) {
+    log_fatal("open(%s): %s", dst, strerror(dst_fd.error()));
+  }
+
+  copy(src_fd->get(), dst_fd->get());
+}
+
+#endif
+
+void remove_backing_files(const std::string &dir, int64_t job_id) {
+  uint8_t group_id = job_id & 0xFF;
+  std::string job_dir = wcl::join_paths(dir, wcl::to_hex(&group_id), std::to_string(job_id));
+  auto dir_range = wcl::directory_range::open(job_dir);
+  if (!dir_range) {
+    log_fatal("opendir(%s): %s", job_dir.c_str(), strerror(dir_range.error()));
+  }
+
+  // loop over the directory now and delete any files as we go.
+  for (const auto &entry : *dir_range) {
+    if (!entry) {
+      log_fatal("readdir(%s): %s", job_dir.c_str(), strerror(entry.error()));
+    }
+    if (entry->type != wcl::file_type::regular) {
+      log_fatal("remove_backing_files(%s): found non-regular entry: %s", job_dir.c_str(),
+                entry->name);
+    }
+    unlink_no_fail(entry->name.c_str());
+  }
+}
+
+void remove_backing_files(std::string dir, const std::vector<int64_t> &job_ids,
+                          size_t min_removals_per_thread, size_t max_number_of_threads) {
+  // Calculate a good number of threads to use.
+  size_t actual_num_threads =
+      std::min(max_number_of_threads, std::max(1UL, job_ids.size() / min_removals_per_thread));
+
+  // Calculate how many removals will be done by each task
+  size_t actual_removals_per_thread =
+      job_ids.size() / actual_num_threads + !!(job_ids.size() % actual_num_threads);
```

Comment:
> `!!(job_ids.size() % actual_num_threads)`

This looks like "add one if job_ids.size isn't an exact multiple of actual_num_threads" but it took some time to figure that out and I'm not even sure that I'm write so this probably warrants a comment

Reply:
Yeah sorry lol. That is in fact what it does. I always try to pair divisions and modulus when I have to use both but its kind of a silly optimization. It basically just ensures that the compiler will put the div and the mod in the same basic block and will find the optimization to emit the single divmod instruction more easily instead of doing two division operations...this is without a doubt a pre-mature and un-tested optimization lol. I've never disassembled things when I did this to see if it even matters lol.